File: src/com/moneydance/modules/features/formula/FormulaWindow.java
/************************************************************\
 *      Copyright (C) 2016 The Infinite Kind, Limited       *
\************************************************************/

package com.moneydance.modules.features.formula;

import com.moneydance.apps.md.view.gui.OKButtonListener;
import com.moneydance.apps.md.view.gui.OKButtonPanel;
import com.moneydance.awt.AwtUtil;
import com.moneydance.awt.GridC;
import com.moneydance.modules.features.formula.reminder.ReminderDetails;
import com.moneydance.modules.features.formula.reminder.ReminderList;
import com.moneydance.util.UiUtil;

import javax.swing.*;
import java.awt.*;
import java.awt.event.WindowEvent;

/** Window used for Account List interface */

public class FormulaWindow
  extends JFrame
  implements OKButtonListener
{
  public FormulaWindow(MDApi api) {
    super("Formulas");

    ReminderDetails reminderDetails = new ReminderDetails(api);

    ReminderList reminderList = new ReminderList(api);
    reminderList.addSelectionListener(reminder ->
      reminderDetails.setReminder(reminder)
    );

    JPanel p = new JPanel(new GridBagLayout());
    p.setBorder(BorderFactory.createEmptyBorder(UiUtil.DLG_VGAP, UiUtil.DLG_HGAP,
            UiUtil.DLG_VGAP, UiUtil.DLG_HGAP));
    p.add(reminderList, GridC.getc(0, 0).wxy(1, 1).fillboth());

    p.add(reminderDetails, GridC.getc(0,1).wxy(1, 1).fillboth());

    p.add(Box.createVerticalStrut(UiUtil.DLG_VGAP), GridC.getc(0,2));

    final OKButtonPanel okPanel = new OKButtonPanel(api.getGUI(), this, OKButtonPanel.QUESTION_OK_CANCEL);
    p.add(okPanel, GridC.getc(0,3).east());
    getContentPane().add(p);

    setDefaultCloseOperation(JFrame.DO_NOTHING_ON_CLOSE);
    enableEvents(WindowEvent.WINDOW_CLOSING);

    setSize(500, 400);
    AwtUtil.centerWindow(this);
  }

  public final void processEvent(AWTEvent evt) {
    if(evt.getID()==WindowEvent.WINDOW_CLOSING) {
      goAway();
      return;
    }
    if(evt.getID()==WindowEvent.WINDOW_OPENED) {
    }
    super.processEvent(evt);
  }

  void goAway() {
    setVisible(false);
    dispose();
  }

  @Override
  public void buttonPressed(int buttonId) {
    if (buttonId == OKButtonPanel.ANSWER_CANCEL) {
      goAway();
    } else if (buttonId == OKButtonPanel.ANSWER_OK) {
      goAway();
    }
  }
}

File: src/main/kotlin/com/moneydance/modules/features/formula/FormulaWindow.kt
package com.moneydance.modules.features.formula

import javax.swing.JFrame
import com.moneydance.apps.md.view.gui.OKButtonListener
import com.moneydance.modules.features.formula.reminder.ReminderList
import com.moneydance.modules.features.formula.reminder.ReminderDetails
import javax.swing.JPanel
import java.awt.GridBagLayout
import javax.swing.BorderFactory
import com.moneydance.util.UiUtil
import com.moneydance.awt.GridC
import com.moneydance.apps.md.view.gui.OKButtonPanel
import java.awt.event.WindowEvent
import com.moneydance.awt.AwtUtil
import java.awt.AWTEvent
import javax.swing.Box

/** Window used for Formula interface  */
class FormulaWindow(api: MDApi) : JFrame("Formulas"), OKButtonListener {
    private val reminderList = ReminderList(api)

    init {
        val reminderDetails = ReminderDetails(api)
        reminderList.addSelectionListener { selected -> reminderDetails.setReminder(selected) }
        val p = JPanel(GridBagLayout())
        p.border = BorderFactory.createEmptyBorder(
            UiUtil.DLG_VGAP, UiUtil.DLG_HGAP,
            UiUtil.DLG_VGAP, UiUtil.DLG_HGAP
        )
        p.add(reminderList, GridC.getc(0, 0).wxy(1f, 1f).fillboth())
        p.add(reminderDetails, GridC.getc(0, 1).wxy(1f, 1f).fillboth())
        p.add(Box.createVerticalStrut(UiUtil.DLG_VGAP), GridC.getc(0, 2))
        val okPanel = OKButtonPanel(api.gui, this, OKButtonPanel.QUESTION_OK)
        p.add(okPanel, GridC.getc(0, 3).east())
        contentPane.add(p)
        defaultCloseOperation = DO_NOTHING_ON_CLOSE
        enableEvents(WindowEvent.WINDOW_CLOSING.toLong())
        setSize(1000, 700)
        AwtUtil.centerWindow(this)
    }

    fun reload() {
        reminderList.reload()
    }

    public override fun processEvent(evt: AWTEvent) {
        if (evt.id == WindowEvent.WINDOW_CLOSING) {
            goAway()
            return
        }
        if (evt.id == WindowEvent.WINDOW_OPENED) {
        }
        super.processEvent(evt)
    }

    fun goAway() {
        isVisible = false
        dispose()
    }

    override fun buttonPressed(buttonId: Int) {
        goAway()
    }
}